pythonequipmentdrivers/utility/data_management.py

```diff
@@ -109,6 +110,92 @@ def dump_data(directory, file_name, data):
     return None
 
 
+def dump_array_data(file_path: Path, data: tuple, init=False,
+                    longest=False, fillvalue=None) -> None:
+    """
+    dump_array_data(file_path, data, init=False, longest=False, fillvalue=None)
+
+    Writes a tuple of iterables to rows of a csv file by transposing them.
+    Useful for logging data contained in multiple arrays, row by row after
+    a test or measurement completes.
+    if longest=True, will pad with fillvalue, default None
+
+    example:
+
+
+    If init=True a new file is created (or overwriten if the file already
+    exists) and the iterable "data" is unpacked into the first row of the new
+    document. If init = False then the existing file is opened and the iterable
+    "data" is unpacked and appended to the end of the document. If init = False
+    but the file does not already exist it will be created.
+
+    The arguement "file_path" specifies the path of the file to be
+    created/updated. The log file will be storted at file_path.csv
+    , file_path itself does not need to contain the file extension, it will
+    automatically be added.
+
+    The iterable "data" can contain an arbitrary number of elements and each
+    element can use an arbitrary data type as long as it can be safely cast to
+    a string. The argument "data" SHOULD be passed as a PACKED structure,
+    i.e. AS an iterable. For example:
+
+        Incorrect:
+            dump_array_data("my_data", 1, 2, 3, 4, 5)
+            dump_array_data("my_data", *list_of_data)
+            dump_array_data("my_data", *['a', 'b', 'c', 4, 5, 6])
+            dump_array_data("my_data", *(1, 2, 3, 4, 5))
+            dump_array_data("my_data", 'column 1', 'column 2', init=True)
+
+        Correct:
+            dump_array_data("my_data", [1, 2, 3, 4, 5])
+            dump_array_data("my_data", (['a', 'b', 'c', 'd'], [1, 2, 3, 4, 5]),
+                            init=True)
+
+    Args:
+        file_name (str, or path-like object): path of the log file to use, does
+            not need to include the file extension or previously exist.
+        data: an iterable of data to be stored.
+
+    Kwargs:
+        init (bool, optional): Whether or not to open the log file in write
+            mode ("True", for creating a new file) or append mode ("False" for
+            adding additional data). Defaults to False.
+    Example:
+        cwd = Path().parent.resolve()
+        file = cwd.joinpath('my_data')
+        list1 = ['a', 'b', 'c', 'd', 'e']
+        list2 = [1, 2, 3, 4, 5, 6, 7, 8]
+        list3 = [4, 5, 6, 7, 8, 9]
+        args = (list1, list2, list3)
+        dump_array_data(file, args, longest=True)
+
+        outputs as:
+        a,1,4
+        b,2,5
+        c,3,6
+        d,4,7
+        e,5,8
+        None,6,9
+        None,7,None
+        None,8,None
+
+    Returns:
+        None
+    """
+
+    mode = 'w' if init else 'a'
```

Comment:
Here's my suggestion:
- get rid of mode (only used once)
- get rid of un-needed return (returning None is specified by type-hinting in the prototype)
- change "file_path.name" to "file_path.stem" to avoid creating files with two extensions
- since "fillvalue" defaults to None you can use that as a flag for whether or not to use longest
    - This eliminates the need for the arg "longest"
    - Also you don't need to remember to pass two arguments when you want to use the fill value
- creating the iterator before using it to write the data to file reduces the level of nesting without increasing the code length
- update type hinting, "fillvalue" is an optional string (Optional is imported from typing), init is a bool, "data" is will actually work for non-tuples I would change it to "Iterable[Iterable[Any]]" as its an iterable of iterables of anything that can be cast to a string (both imported from typing).
- change the name of "vals" to be more explicit (represents a row of data in a csv)
```
def dump_array_data(file_path: Path,
                    data: Iterable[Iterable[Any]],
                    init: bool = False,
                    fill_value: Optional[str] = None) -> None:

    file_path = Path(file_path)
    file_path_ext = file_path.parent / f'{file_path.stem}.csv'
    
    if fill_value is not None:  # create generator 
        rows = zip_longest(*data, fillvalue=fill_value)
    else:
        rows = zip(*data)

    with open(file_path_ext, 'w' if init else 'a') as f:
            for row in values:
                print(*row, sep=',', end='\n', file=f)
```